cloud/clouds.go

```diff
@@ -134,28 +173,91 @@ func PublicCloudMetadata(searchPath ...string) (result map[string]Cloud, fallbac
 	return clouds, true, err
 }
 
+var logger = loggo.GetLogger("juju.cloud")
+
 // ParseCloudMetadata parses the given yaml bytes into Clouds metadata.
 func ParseCloudMetadata(data []byte) (map[string]Cloud, error) {
-	var metadata clouds
-	err := yaml.Unmarshal(data, &metadata)
-	if err != nil {
+	var metadata cloudSet
+	if err := yaml.Unmarshal(data, &metadata); err != nil {
 		return nil, errors.Annotate(err, "cannot unmarshal yaml cloud metadata")
 	}
-	metadata.denormaliseMetadata()
-	return metadata.Clouds, nil
+
+	// Translate to the exported type. For each cloud, we store
+	// the first region for the cloud as its default region.
+	clouds := make(map[string]Cloud)
+	for name, cloud := range metadata.Clouds {
+		var regions []Region
+		if len(cloud.Regions.Map) > 0 {
+			logger.Debugf("%v", cloud.Regions)
+			for _, item := range cloud.Regions.Slice {
+				name := fmt.Sprint(item.Key)
+				r := cloud.Regions.Map[name]
```

Comment:
This means I don't have to extract map items explicitly, so easier to maintain (Region may grow fields). The MapSlice is used only for the key ordering.